src/main/java/randoop/generation/Bloodhound.java

```diff
@@ -0,0 +1,210 @@
+
+    // Default weight is uniform probability.  If is used if no coverage details are available.
+    // This is the case for classes like java.lang.Object (not explicitly under test).
+    double weight = 1.0 / numOperations;
+
+    // Recompute weights for all operations.
+    for (TypedOperation operation : operationSimpleList) {
+      CoverageTracker.BranchCoverage covDet =
+          CoverageTracker.instance.getDetailsForMethod(operation.getName());
+
+      if (covDet != null) {
+        // Coverage details are available.
+
+        // The number of successful invocations of this method. Corresponds to succ(m).
+        Integer numSuccessfulInvocations = methodSuccessfulCalls.get(operation);
+        if (numSuccessfulInvocations == null) {
+          numSuccessfulInvocations = 0;
+        }
+
+        // If this method has no uncovered branches, or it is the most-invoked method,
+        // use the default weight and skip this step.
+        if (covDet.uncoveredBranches != 0 && numSuccessfulInvocations != maxSuccessfulCalls) {
+          // Uncovered branch ratio of this method. The name uncovRatio(m) is from the GRT paper.
+          double uncovRatio;
+          if (covDet.totalBranches == 0) {
+            uncovRatio = 0.5;
+          } else {
+            uncovRatio = (double) covDet.uncoveredBranches / covDet.totalBranches;
+          }
+
+          // Call ratio of this method. Corresponds to succ(m) / maxSucc(M) in the GRT paper.
+          double callRatio;
+
+          if (maxSuccessfulCalls == 0) {
+            callRatio = 0.5;
+          } else {
+            callRatio = numSuccessfulInvocations.doubleValue() / maxSuccessfulCalls;
+          }
+
+          // Corresponds to w(m, 0) in the GRT paper.
+          weight = alpha * uncovRatio + (1 - alpha) * (1 - callRatio);
+
+          // Corresponds to the k variable in the GRT paper.
+          Integer numSelectionsOfMethod = methodSelections.get(operation);
+          if (numSelectionsOfMethod != null) {
+            // Corresponds to the case where k >= 1 in the GRT paper.
+            double val1 =
+                (-3.0 / Math.log(1 - p))
+                    * (Math.pow(p, numSelectionsOfMethod) / numSelectionsOfMethod);
+            double val2 = 1.0 / Math.log(numOperations + 3);
+            weight *= Math.max(val1, val2);
+          }
+        }
+      }
+
+      methodWeights.put(operation, weight);
+    }
+  }
+
+  /**
+   * First, update the weights of all methods under test. Retrieve the next method for constructing
+   * a new sequence while also considering each method's weights. Update the number of times the
+   * method has been selected.
+   *
+   * @return the chosen {@code TypedOperation} for the new sequence
+   */
+  @Override
+  public TypedOperation selectOperation() {
+    // Collect branch coverage and recompute weights for methods under test.
+    updateBranchCoverageMaybe();
+    updateWeightsForOperations();
```

Comment:
Why is updateWeightsForOperations called on every step?  It's called in selectOperation, right after updateBranchCoverageMaybe which only has an effect sometimes.
My first concern was the efficiency cost of re-computing weights every operation.
Also, the GRT paper says, "At each update of the coverage, the weights are recalculated, and k is reset to 0."  This implies that weights aren't recomputed at every step.

Should you replace

```
    updateBranchCoverageMaybe();
    updateWeightsForOperations();
```

by

```
    stepNum++;
    if (stepNum % branchCoverageInterval == 0) {
      updateBranchCoverage();
      updateWeightsForOperations();
    }
```

?

Reply:
Hi Mike, 
I'm wondering if I'm misinterpreting this segment, but I do believe that weights are regularly updated at each step.
![image](https://user-images.githubusercontent.com/15660492/38590653-6fa7efd6-3ce7-11e8-95dc-970c59da2143.png)
In this segment, the authors define `w` as a function of the method and `k`. `k` is the "the number of selections of method m since the last update of the coverage information". Thus, it appears to be the case that a method's weight will change within the middle of an interval (where coverage is computed at the ends of the interval). 

In regards to efficiency, at the moment, I've implemented it such that the weights of ALL methods under test are recomputed at every step. I wonder if it's sufficient to only update the weight of the method that gets chosen to be used to create a "new and unique sequence".

Reply:
And then, at each point where coverage information is updated/re-summarized, we would then recompute the weights of all the methods under test.

Reply:
Thanks for the clarification.  I had interpreted k as the number of selections of method m between updates to coverage information (that is, between the previous two coverage updates).  Your interpretation may be right -- it fits the text even better.

Updating only the weight of the method that gets chosen is sufficient and probably cheaper, since there are many methods under test.  (But, I haven't run timing tests.)  Could you try that?